Out-of-Vocab for VALUES is not an error

joka921 opened the pull request:
This fixes #290.

When an out-of-vocab entity is used in a `VALUES` clause this previously triggered an exception which is wrong. This PR now ignores the affected rows and emits warnings that are passed to the user in the
`Json` response.

Internally this implements a general mechanism to transport warnings for all kinds of operations.

hannahbast commented on test/StringSortComparatorTest.cpp:
```diff
@@ -41,6 +41,22 @@ TEST(LocaleManagerTest, Punctuation) {
 
+TEST(LocaleManagerTest, Normalization) {
+  // é as single codepoints
+  const char a[]  = {static_cast<char>(0xC3), static_cast<char>(0xA9), 0 };
```
Simpler: ... = "\xc3\xa9"

hannahbast commented on test/StringSortComparatorTest.cpp:
```diff
@@ -41,6 +41,22 @@ TEST(LocaleManagerTest, Punctuation) {
+  // é as single codepoints
+  const char a[]  = {static_cast<char>(0xC3), static_cast<char>(0xA9), 0 };
+  // é as e + accent aigu
+  const char b[] = {'e', static_cast<char>(0xCC), static_cast<char>(0x81), 0};
```
Dito: ... = "e\xcc\x81"

hannahbast commented on src/engine/Server.cpp:
```diff
@@ -335,47 +335,39 @@ string Server::composeResponseJson(const ParsedQuery& query,
+    _requestProcessingTimer.cont();
+    qet.writeResultToStreamAsJson(os, query._selectedVariables, limit, offset,
+                                  maxSend);
+    j["res"] = json::parse(os.str());
```
Is this efficient for a huge result (e.g. millions or billions of triples)?

However, the UI only asks for the Top-X results and big results are usually exported as CSV or TSV

hannahbast commented on src/engine/Server.cpp:
```diff
@@ -335,47 +335,39 @@ string Server::composeResponseJson(const ParsedQuery& query,
+  j["time"]["computeResult"] = std::to_string(compResultUsecs / 1000.0) + "ms";
 
-  return os.str();
+  return j.dump(4);
```
See the comment above

hannahbast commented on src/engine/Values.cpp:
```diff
@@ -138,16 +138,28 @@ void Values::writeValues(IdTable* res, const Index& index,
+      result(numActuallyWritten, colIdx) = id;
     }
+    numActuallyWritten++;
+  skipRow:;
```
Is the indentation correct for this label?

joka921 replied:
Yes

hannahbast commented on src/engine/Values.cpp:
```diff
@@ -138,16 +138,28 @@ void Values::writeValues(IdTable* res, const Index& index,
+        getWarnings().push_back("The word " + row[colIdx] +
+                                " is not part of the vocabulary.");
+        numSkipped++;
+        goto skipRow;
```
Please add a comment so that it's clear that this is used simply to break out of the inner loop (and nothing worse) and that one finds the corresponding label a few lines further down

hannahbast commented on test/StringSortComparatorTest.cpp:
```diff
@@ -41,6 +41,22 @@ TEST(LocaleManagerTest, Punctuation) {
 
+TEST(LocaleManagerTest, Normalization) {
+  // é as single codepoints
+  const char a[] = {static_cast<char>(0xC3), static_cast<char>(0xA9), 0};
```
Simplify to: ... = "\xc3\xa9"

joka921 commented:
Ok, The ICU-Level discussion goes on: @hannahbast @niklas88 
After rethinking this again I found the following: We need the `IDENTICAL` Level always.
As soon as there can be words, that compare Equal but are not our vocabulary breaks apart. This is a bug that does not often occur in "typical" queries but it is a bug nonetheless.
(we strictly need to consecutive entries in the vocabulary to compare not equal. I am even thinking about enforcing it by additionally running `strcmp` if the Unicode collation returns equal. That way we might save the `Quarternary` Level (I just got the idea while writing this). But I also first want to discuss this.

Another thing that is already open for review are the changes in `QueryExecutionTree.h/.cpp` and `Server.cpp` where I refactored the Json pipeline to consistently use nlohmann::json.

hannahbast commented:
@joka921 @niklas88 Thanks, Johannes! Can you be more specific with what you mean by "our vocabulary breaks apart", what exactly happens?

joka921 commented:
Ok so now i have 
- Adressed the small wishes from @hannahbast 's review.
- Removed all Unicode-fixing business out of this PR and refactored it to another branch since those
  are completely different issues.

niklas88 commented on e2e/queryit.py:
```diff
@@ -147,6 +147,18 @@ def test_check(check_dict: Dict[str, Any], result: Dict[str, Any]) -> bool:
+            for requested_warning in value:
+                found = False
+                for actual_warning in result["warnings"]:
+                    if actual_warning.startswith(requested_warning):
```
`.startswith()` is already a bit fuzzy, why not just do `.contains()` and then it is all about containing a warning

niklas88 commented on src/engine/QueryExecutionTree.h:
```diff
@@ -145,144 +147,22 @@ class QueryExecutionTree {
   size_t _sizeEstimate;
 
   std::shared_ptr<const ResultTable> _cachedResult = nullptr;
-  void writeJsonTable(
```
Yay, no more manual JSON building, very good!

niklas88 commented:
@hannahbast have your questions/requested changes been answered, I think this can be merged
